File: src/main/java/io/github/chaosdave34/kitpvp/abilities/impl/creeper/ExplodeAbility.java
package io.github.chaosdave34.kitpvp.abilities.impl.creeper;

import io.github.chaosdave34.ghutils.persistentdatatypes.StringArrayPersistentDataType;
import io.github.chaosdave34.kitpvp.KitPvp;
import io.github.chaosdave34.kitpvp.abilities.Ability;

import net.kyori.adventure.text.Component;
import org.bukkit.NamespacedKey;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.entity.EntityExplodeEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.persistence.PersistentDataContainer;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class ExplodeAbility extends Ability {
    public ExplodeAbility() {
        super("explode", "Explode", Type.SNEAK, 25);
    }

    @Override
    public @NotNull List<Component> getDescription() {
        return createSimpleDescription("Cause an explosion.");
    }

    @Override
    public boolean onAbility(Player p) {
        p.getLocation().createExplosion(p, 3, false);
        return true;
    }

    @EventHandler
    public void onExplode(EntityExplodeEvent e) {
        if (e.getEntity() instanceof Player p) {
            for (ItemStack armorContent : p.getInventory().getArmorContents()) {
                if (armorContent == null) continue;

                PersistentDataContainer container = armorContent.getItemMeta().getPersistentDataContainer();
                NamespacedKey key = new NamespacedKey(KitPvp.INSTANCE, "abilities");

                if (container.has(key)) {
                    String[] abilities = container.get(key, new StringArrayPersistentDataType());
                    if (abilities == null) return;

                    for (String id : abilities) {
                        if (getId().equals(id)) {
                            e.blockList().removeIf(block -> !block.hasMetadata("placed_by_player"));
                            return;
                        }
                    }

                }
            }
        }
    }
}

File: src/main/kotlin/io/github/chaosdave34/kitpvp/abilities/impl/ExplodeAbility.kt
package io.github.chaosdave34.kitpvp.abilities.impl

import io.github.chaosdave34.kitpvp.abilities.Ability
import net.kyori.adventure.text.Component
import org.bukkit.block.Block
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.entity.EntityExplodeEvent

class ExplodeAbility : Ability("explode", "Explode", Type.SNEAK, 15) {

    override fun getDescription(): List<Component> = createSimpleDescriptionAsList("Cause an explosion.")

    override fun onAbility(player: Player): Boolean {
        var power = 2f
        if (player.isGlowing) {
            player.isGlowing = false
            power *= 2
        }
        player.location.createExplosion(player, power, false)
        return true
    }

    @EventHandler
    fun onExplode(event: EntityExplodeEvent) {
        val player = event.entity
        if (player is Player) {
            for (armorContent in player.inventory.armorContents) {
                if (armorContent?.hasThisAbility() == true) {
                    event.blockList().removeIf { block: Block -> !block.hasMetadata("placed_by_player") }
                    return
                }
            }
        }
    }
}